compiler/qsc_parse/src/item.rs

```diff
@@ -131,14 +132,74 @@ fn parse_top_level_node(s: &mut ParserContext) -> Result<TopLevelNode> {
         Ok(TopLevelNode::Stmt(stmt))
     }
 }
+pub fn parse_implicit_namespace(source_name: &str, s: &mut ParserContext) -> Result<Namespace> {
+    let lo = s.peek().span.lo;
+    let items = parse_namespace_block_contents(s)?;
+    let span = s.span(lo);
+    let namespace_name = source_name_to_namespace_name(source_name, span)?;
+
+    Ok(Namespace {
+        id: NodeId::default(),
+        span,
+        doc: "".into(),
+        name: namespace_name,
+        items: items.into_boxed_slice(),
+    })
+}
+
+/// Given a file name, convert it to a namespace name.
+/// For example, `foo/bar.qs` becomes `foo.bar`.
+fn source_name_to_namespace_name(raw: &str, error_span: Span) -> Result<Idents> {
+    let path = std::path::Path::new(raw);
+    let mut namespace = Vec::new();
+    for component in path.components() {
+        match component {
+            std::path::Component::Normal(name) => {
+                // strip the extension off, if there is one
+                let mut name = name
+                    .to_str()
+                    .ok_or(Error(ErrorKind::InvalidFileName(error_span)))?;
+
+                if let Some(dot) = name.rfind('.') {
+                    name = name[..dot].into();
+                }
+                // verify that the component only contains alphanumeric characters, and doesn't start with a number
+
+                validate_namespace_name(error_span, name)?;
```

Comment:
Since you're parsing here anyway, why not use the `Ident` created in this function instead of creating a new one?